Kernel Ridge Classification

cperales opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

Kernel Ridge Regression (the only Kernel Ridge estimator is right now implemented in master) has a real simple mathematical formulation to adjust a regression problem using squared error loss combined with l2 regularization. It is similar in SVM is done, but with equal constrictions instead of inequal ones as in SVM.

![elm_min](https://user-images.githubusercontent.com/17336458/35975459-276dd2ec-0cdd-11e8-9256-af6d894dce7c.png)

Using a vector encoding, this regressor can be adapted as a multilabel classifier while keep all the advantages over SVM. And this is my proposal. I followed the mathematical explaining from section 3 in [Venkatesh's paper from 2007](http://dro.deakin.edu.au/eserv/DU:30044591/venkatesh-facerecognition-2007.pdf). Basically, it transforms the target with *r* labels into a vector with *r* components.

#### Any other comments?

This pull request comes after a controversial one about Kernel Extreme Learning Machine ( https://github.com/scikit-learn/scikit-learn/pull/10602 ). As @amueller pointed, ELM is so similar to Kernel Ridge that it is possibly a scam because it [Huang presents in 2012](http://www.neuromorphs.net/nm/raw-attachment/wiki/2015/scc15/ELM-Unified-Learning.pdf) ELM as a new algorithm. But Kernel Ridge, as it is presented in `scikit-learn`, just allows regression. This is a modification, using `sklearn.kernel_ridge.KernelRidge` as base, to allow Kernel Ridge multilabel Classification. Name, tests, code and implementation is open to discussion.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Kernel Extreme Learning Machine added

Commit:
Travis to run tests

Commit:
RBF test for ELM

Custom kernel test for ELM

Commit:
.travis.yml back to original

Commit:
PEP8 modifications

Commit:
Test fixed

TomDLT committed:
[MRG+1] TSNE: only compute error when needed (#10593)

FrankHui committed:
DOC/COSMIT update sgd_fast.pyx (#10601)

1. line 159, remove redundant parentheses 
2. line 399-400, add intercept_decay to docstring

qinhanmin2014 committed:
[MRG+1] Deprecate axis parameter in imputer (#10558)

jnothman committed:
COSMIT Use comb2(x) instead of comb(x, 2) (#10606)

Johayon committed:
[MRG+1] change threshold computation in trees (#10536)

gxyd committed:
[MRG + 1] KMeans optimisation for array C/F contiguity (#10471)

TobiasMadsen committed:
[MRG+1] Allow vector pseudocounts Multinomial NB (#10346) (#10350)

jnothman committed:
[MRG+1] TST apply transformer checks to any estimator supporting 'transform' (#10474)

Commit:
Test for KELM were revised. Also, .update to dicts were removed, for compatibility with Python 2

test_precomputed fixed

test_tweak_params fixed

test_bad_input fixed

Unicode test fixed

test_linear_elm fixed

test_linear_elm_iris fixed

Documentation fixed

unused imports removed

Unused test was removed

Problem with __init__ solved

Commit:
PEP corrections with flake8

Commit:
Code adapted to Python 2 and 3
Float correction

Unicode correction

PEP corrections with flake8

Unicode treatment for Python 2 and Python 3

Commit:
Compatibility for Python 2 and 3 in unicode, while passing flake8 test

Compatibility for Python 2 and 3 in unicode, while passing flake8 test

jnothman committed:
MAINT hide test of deprecated code that is failing on some arch (#10479)

Commit:
Try/Except for unicode handling changed to a more pythonic way

Commit:
Merge branch 'master' into KELM

Commit:
Kernel Extreme Learning Machine implemented as a Kernel Ridge Classification

Commit:
Objects and tests adapted

Commit:
Tests passed

Commit:
More tests works, and flake8 too

Commit:
Flake8 correction

Commit:
Documentation

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -0,0 +1,165 @@
+            # Check alpha
+            if self.alpha <= 0:
+                raise ValueError('C must be positive')
+            X, y = check_X_y(X, y)
```
so multi-output isn't an option?

cperales replied:
The QP problem tries to minimize both l2 norm of the errors and l2 norm of the dual coeficient. 

Alpha multiplies the norm of the matrix, so this parameter is a measure of how important is the norm of the matrix with respect to the norm of the errors. What it is wrong is *C must be positive*, is *alpha* what it should appear in that message. Also, reasoning is both valid for regression and classification, so I'll move above condition.

What do you refer about with multi-output?

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -0,0 +1,165 @@
+            n = X.shape[0]
+            if self.classes_.shape[0] == 1:
+                self.n_classes_ = int(self.classes_[0] + 1)
+                self.classes_ = np.arange(self.n_classes_)
```
self.classes_ needs to encode all that was in `y` in the input. It's your job as the estimator developer to convert back to that, not to alter what the user provides.

cperales replied:
You're right, this will be fixed in the next commit

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -0,0 +1,165 @@
+                self.n_classes_ = len(self.classes_)
+
+            T = np.zeros((n, self.n_classes_), dtype=np.float64)
+            # It is essential in order to adapt it to string labels
```
Why don't you use a LabelEncoder like other classifiers do?

cperales replied:
I didn't know about that, but it seems so easy to use. I think the label encoder it is needed is `sklearn.preprocessing.LabelBinarizer`

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -0,0 +1,165 @@
+    def _pairwise(self):
+        return self.kernel == "precomputed"
+
+    def fit(self, X, y=None, sample_weight=None):
```
not a lot of fit seems to be shared between the regressor and classifier. why not just put each fit into its corresponding class.

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,228 @@
+    X = "foo!"  # input validation not required when KRC is not fitted
+
+    clf = KRC()
+    assert_raises_regexp(Exception, r"not\b.*\bfitted\b",
```
This kind of thing is tested by common tests. You do not need to.

cperales replied:
OK then.

Note: I just added because, due to similarity between KRC and SVC, I saw may tests from `sklearn.svm.tests.test_svm.py` could work also for KRC (and they do, as expected).

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,228 @@
+        clf = KRC(kernel=k).fit(iris.data, iris.target)
+        assert_greater(np.mean(clf.predict(iris.data) == iris.target), 0.75)
+
+    assert_array_equal(clf.classes_, np.sort(clf.classes_))
```
This kind of thing is tested by common tests. You do not need to.

cperales replied:
Same that ( https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168151047 ), I'll remove it.

jnothman reviewed (commented):
Thanks for the PR. You should expect review to take a while though I admit that the estimator does not include a lot of new code, so perhaps try to share code with the existing tests (e.g. using pytest.mark.parametrize to run on both KRR and KRC) to minimise the workload for reviewers. 

Your examples files are named inappropriately now.

We still need just KernelRidge to be the regressor (and not KRR and not KRC and not KernelRidgeRegressor)

KernelRidgeClassifier needs to be added to doc/modules/classes.rst

Conventionally, a module which shares a base implementation with other modules is called base.py, and the class Base*. Please try to follow such conventions.

Commit:
Merge branch 'master' into KernelRidgeClassification

Commit:
Problem with documentation

cperales commented:
Thanks for the revision, I'll try to answer each aspect separately, but first of all, thanks for answering with so many details.

Commit:
Corrections from @jnothman

test_unfitted removed as @jnothman proposed

test_iris removed as @jnothman proposed

LabelBinarized applied

Commit:
alpha raise error moved to classification

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-8a66741a45d0beada1ab7d5ff28bcb4e7aa7b7bc)

**new alerts:**

* 1 for Conflicting attributes in base classes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
PEP mistakes fixed

Try/except remove, it can work with sparse data

Plot with KRC

Tests fixed

sklearn-lgtm commented:
This pull request introduces 1 alert and fixes 5 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-71e107b5fe0a31bc0217ab4b703147a29d20d5f6)

**new alerts:**

* 1 for Conflicting attributes in base classes

**fixed alerts:**

* 2 for Wrong number of arguments for format
* 1 for Variable defined multiple times
* 1 for Nested loops with same variable
* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

cperales committed:
__init__.py removed from example folders
README added to kernel ridge estimators

cperales committed:
sparse removed

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-36ea64e980c0289b5401443b341fa83a8277be3f)

**new alerts:**

* 1 for Conflicting attributes in base classes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
Merge branch 'master' into KernelRidgeClassification

Commit:
kernel_ridge folder added to setup.py

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-2a5805e3ab74d1931860b146b70814f7ad6632e6)

**new alerts:**

* 1 for Conflicting attributes in base classes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
Problem with inheritance solved.
When you have a multiple inheritance, python take first inherited object as main

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-0d3165d0bc6b8aa92e288cdd5ed18255a8eec6d2)

**new alerts:**

* 1 for Conflicting attributes in base classes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
KernelRidgeBase does not inherite from ClassifierMixin or RegressorMixin

Commit:
Merge branch 'master' into KernelRidgeClassification

jnothman commented:
we've had problems with slow travis lately. I think it may be independent
of the pytest issue, but certainly we should start looking into that

cperales commented:
Ah, OK then. I thought I have been doing something bad. Besides, I think the "reorganization" of what is Kernel Ridge Regression and Kernel Ridge Classification could be made better, that's why I made a last commit.

But, if everything goes as expected, your revisions are done and the tests pass, so I expect your answer @jnothman . If I could help you in any way, no doubt to tell me

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +123,41 @@ def fit(self, X, y=None, sample_weight=None):
+                y_ = y_.reshape(-1, 1)
+                ravel = True
+        else:  # Multilabel classification
+            X, y = check_X_y(X, y)
```
Do you mean multiclass rather than multilabel? For multilabel, you need multi_output=True I think

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -2,39 +2,21 @@
-    (i.e., when y is a 2d-array of shape [n_samples, n_targets]).
-
-    Read more in the :ref:`User Guide <kernel_ridge>`.
+class KernelRidgeBase(BaseEstimator):
```
BaseKernelRidge, please

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,86 @@
+from .kernel_ridge import KernelRidgeBase
+
+
+class KernelRidgeClassification(ClassifierMixin, KernelRidgeBase):
```
KernelRidgeClassifier, please

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/__init__.py:
```diff
@@ -0,0 +1,19 @@
+KRR = KernelRidgeRegression
+KRC = KernelRidgeClassification
+
+__all__ = ["KernelRidge",
```
Just KernelRidge and KernelRidgeClassifier please

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/__init__.py:
```diff
@@ -0,0 +1,19 @@
+from .krr import KernelRidgeRegression
+
+KernelRidge = KernelRidgeRegression
+KRR = KernelRidgeRegression
```
remove this please

jnothman replied:
We don't generally have aliases like this

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/krr.py:
```diff
@@ -0,0 +1,104 @@
+from .kernel_ridge import KernelRidgeBase
+
+
+class KernelRidgeRegression(RegressorMixin, KernelRidgeBase):
```
This is currently called KernelRidge and should stay that way

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -69,7 +51,7 @@ class KernelRidge(BaseEstimator, RegressorMixin):
     Attributes
     ----------
-    dual_coef_ : array, shape = [n_samples] or [n_samples, n_targets]
+    F : array, shape = [n_samples] or [n_samples, n_targets]
```
what is this? I can't see where you set F. You shouldn't be dropping `dual_coef_` either.

jnothman replied:
On the other hand, we don't really need a docstring in the base class

cperales replied:
Sorry about F/dual_coef_, I don't know why that was there. Now it's fixed.

I added the docstring because:
- When I was running the tests, some of them failed because this object had `def __init__` but no docstring, so I had to added.
- I think it is useful and it does not disturb.

But if you are really concerned about this, I can remove it. But I think there is one test that would complain.

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -81,27 +63,25 @@ class KernelRidge(BaseEstimator, RegressorMixin):
         Linear ridge regression.
     sklearn.svm.SVR:
         Support Vector Regression implemented using libsvm.
+    sklearn.svm.SVC:
```
I don't think this is relevant, but you should add KernelRidgeClassifier to KernelRidge's see also

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,86 @@
+
+    Attributes
+    ----------
+    F : array, shape = [n_samples] or [n_samples, n_targets]
```
I think this is out of date

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -81,27 +63,25 @@ class KernelRidge(BaseEstimator, RegressorMixin):
       "Machine Learning: A Probabilistic Perspective", The MIT Press
       chapter 14.4.3, pp. 492-493
 
+    * An, Senjian, Wanquan Liu, and Svetha Venkatesh.
```
This should only be in the KernelRidgeClassifier docstring.

But if you drop the BaseKernelRidge docstring, that would help you tease things apart

cperales replied:
Done, that paper now only appear in KernelRidgeclassifier,

About docstring, as I said: if you are really concerned about this, I can remove it, but I think there is one test that would complain.

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -81,27 +63,25 @@ class KernelRidge(BaseEstimator, RegressorMixin):
     """
-    def __init__(self, alpha=1, kernel="linear", gamma=None, degree=3, coef0=1,
+    # By default, it works as a regressor
+    regression = True
```
I don't think we need a default. Just put this in the regressor (or don't, which is my preference)

cperales replied:
Replied in your review about `_estimator_type`

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +123,41 @@ def fit(self, X, y=None, sample_weight=None):
-        self.dual_coef_ = _solve_cholesky_kernel(K, y, alpha,
+
+        ravel = False
+        if self.regression is True:
```
just use _estimator_type? or the is_classifier function?

jnothman replied:
(And usually we wouldn't bother with `is True`)

cperales replied:
Totally righ, `regression` property can be removed now I understand better inheritances from `ClassifierMixin` and `RegressionMixin`.

However, some stylo tests complains if I ask for a property that does not appear, so I instanced estimator_type = None` in `BaseKernelRidge`.

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,86 @@
+
+    X_fit_ : {array-like, sparse matrix}, shape = [n_samples, n_features]
+        Training data, which is also required for prediction
+
```
classes_ too

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -81,27 +63,25 @@ class KernelRidge(BaseEstimator, RegressorMixin):
+    # By default, it works as a regressor
+    regression = True
+
+    def __init__(self, alpha=1, kernel="rbf", gamma=None, degree=3, coef0=1,
```
We can't change default values from version to version. Leave the default kernel as 'linear', at least in KRR.

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -181,5 +179,12 @@ def predict(self, X):
+
+        if self.regression is False:  # Multilabel classification
+            # Decoding
+            y = self.label_encoder_.inverse_transform(y)
```
Don't we need to do an argmax before this? How is this not failing tests? Am I missing something?

cperales replied:
As far as I tested with `LabelBinarized`, that is its work. Before you tell me about that class, I had to do that work, but now `LabelBinarized` does the job

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,203 @@
+iris.target = iris.target[perm]
+
+
+def test_check_estimator():
```
This should be run automatically in sklearn/tests/test_common.py:def test_non_meta_estimators

jnothman replied:
You could however run `check_estimator(KernelRidgeClassifier(metric='precomputed'))`

cperales replied:
ok, so, test removed. Also, there already is a test for precomputed kernels.

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +123,41 @@ def fit(self, X, y=None, sample_weight=None):
+
+            # This value is similar to number of neurons in hidden
+            # layer in neural version of Extreme Learning Machine
+            self.h_ = K.shape[1]
```
Isn't `self.h_ == self.fit_X_.shape[0]`?? Why do we need to store it separately? Also, `h_` is a terrible name.

cperales replied:
Sorry, that comes from my previous pull request about ELM, because this value was the equivalent to the number of neurons from the hidden layer in neural network version of Extreme Learning Machine... but more and more it does not make sense to me. I'll remove it

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,203 @@
+    assert_array_equal(clf.predict(X), Y)
+
+
+def test_precomputed():
```
Surely we can share a test like this between KRR and KRC, without duplicating the code. this is also not very easy to read as a test -- I'm sure it could be simpler, but don't see the point in detailed comments until it is refactored with respect to KKR tests.

cperales replied:
Well, as far as I have learned from sklearn tests, a lot of tests are focus to the estimator you are testing, instead of the object itself.

So, Kernel Ridge Regressor and Kernel Ridge Classifier are tested in different ways, because, although they share same mathematical concepts and formulations, first one is used for predicting continue targets and second one for a discrete target, a number of classes.

jnothman replied:
Generic estimator API is tested in common tests.

Tests are often written like some kind of proof by induction. You show a basic feature - like the ability to classify - works, then you show that modifying some parameter works too. Having shown that you can predict from `dual_coef_`, showing that `dual_coef_` is calculated correctly as a function of `kernel` is  sufficient to suggest that all will work together.

cperales replied:
Hmm, I see.. So what you are saying is that some tests are superfluous, because same property is tested twice, one as a regressor and one as a classifier, right?

I'll take a look today and I'll try to reduce the number of tests then.

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -181,5 +179,12 @@ def predict(self, X):
+        # Input validation
         K = self._get_kernel(X, self.X_fit_)
-        return np.dot(K, self.dual_coef_)
+        y = np.dot(K, self.dual_coef_)
```
I suspect KernelRidgeClassifier should also have a `decision_function` reporting this

cperales replied:
I don't understand very well your comment.
I just write that, in case of working as a regressor, that's it, but working as a classifier, a inverse transformation (from encoded labels to original labels, using `LabelBinarizer`)

jnothman replied:
Multiclass (linear) classifiers often provide `decision_function` which reports a score for each class for each sample, such that the argmax is the predicted class.

cperales replied:
Ah, I see, for cross validation, right? I'll take a look about how it is implemented in other classifiers

cperales replied:
Ok, `decision_function` was easy to implement, because it is the first part of the prediction in multiclass problems. Then, as you pointed, something like argmax is applied, but through `LabelBinarizer`.

jnothman reviewed (commented):
A quick partial review.

I don't really think this needs to be split into multiple files, but we have no consistent convention regarding that. Whether or not you split it, some tests should be shared between KRR and KRC.

Commit:
Multilabel was a wrong notation, it is multiclass

Commit:
KernelRidgeBase renamed to BaseKernelRidge

Commit:
KernelRidgeClassification renamed to KernelRidgeClassifier

Commit:
KernelRidgeRegression renamed to KernelRidgeRegressor

Commit:
KernelRidgeRegressor renamed to KernelRidge

Commit:
F does not exist, dual_coef_ instead

Commit:
Docstring modified, __init__ default changed, regression replaced by _estimator_type

Commit:
Test KRR works

Commit:
KernelRidge returned to previous default __init__ parameters.
KernelRidgeClassifier has its own __init__

Commit:
classes_ added to docstring in KernelRidgeClassification

Commit:
PEP8 modifications

Commit:
BaseKernelRidge does not inherit from BaseEstimator

sklearn-lgtm commented:
This pull request introduces 6 alerts - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-7601a7f6501a0fa86fd8f81168f2784c92bfe209)

**new alerts:**

* 3 for Comparison using is when operands support __eq__
* 2 for Conflicting attributes in base classes
* 1 for Missing call to __init__ during object initialization

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
removing _estimator_type from BaseKernelRidge

Commit:
is changed to == in order to avoid lgtm alerts

sklearn-lgtm commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-d28396d5ae0d0942cc55a77712998be7c36907fe)

**new alerts:**

* 1 for Missing call to __init__ during object initialization

---

*Comment posted by [lgtm.com](https://lgtm.com)*

cperales commented:
@jnothman Your reviews were applied, except:
- https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168347410 
- https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168347190 
- https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168344916 
- https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168343790 

The no-changes are explained in cited discussion. Essentially, I just keep some code due to what I learn about scikit-learn tests by trial and error.

jnothman commented:
I've not looked through the changes in detail, but if you rename
BaseKernelRidge to _BaseKernelRidge, which I'd recommend, it won't be
automatically tested.

Commit:
Test renamed

cperales commented:
@jnothman (about https://github.com/scikit-learn/scikit-learn/pull/10633#issuecomment-366121451 ) yesterday I learned that, if you don't inherite from `BaseEstimator`, test is not runed. So `KernelRidge` and `KernelRidgeClassifier` now inherit from `BaseEstimator`, but `BaseKernelRidge` doesn't. Besides, it seemed logic to me.

Anyways, I'll rename `BaseKernelRidge` to `_BaseKernelRidge`.

Commit:
BaseKernelRidge renamed to _BaseKernelRidge

Commit:
lgtm alert solved
KernelRidgeClassifier use init from _BaseKernelRidge through super

Default kernel is rbf for KernelRidgeClassifier

Commit:
PEP8 corrections

Commit:
decision_function implemented, along with a test

Commit:
decision_function is a method related to classifiers, not regressors, so it was moved from _BaseKernelRidge to KernelRidgeClassifier.
Besides, doctest was added

Commit:
Ravel applied to 2-classes classification (decision_function removed, it will be included)

Commit:
PEP8 corrections to f642734

Commit:
decision_function is tested with decision.ravel() > 0, so LabelBinarizer should have neg_label = -1

cperales commented:
@jnothman  `decision_function` is now implemented, which led me to understand better `LaberBinarizer`, and how it works. That objects seems really useful, it will avoid me headaches if I have label encoding, machine learning and python all together.

Commit:
Fewer tests

test_kernel removed from kernel_ridge

test_precomputed removed, this part is already tested

test_kernel_ridge_sample_weights is already tested by generic tests

kernel and iris tests removed

Commit:
PEP8 corrections to b0377b2

cperales commented:
Tests were simplified (as @jnothman suggested in https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168347190), so all the revisions and suggestions you made were applied, and the pull request now passes all the tests

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +114,42 @@ def fit(self, X, y=None, sample_weight=None):
+                y_ = y_.reshape(-1, 1)
+                ravel = True
+        elif self._estimator_type == "classifier":  # Multiclass classification
+            X, y = check_X_y(X, y)
```
Why do we not accept sparse?

cperales replied:
I didn't understand them pretty well, but I'm going to change the `check_X_y`. Should I prepare a test with sparse data, or common tests check it? Is there a list of what common tests check and what not?

cperales replied:
Done

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +114,42 @@ def fit(self, X, y=None, sample_weight=None):
+            if self.alpha <= 0:
+                raise ValueError('alpha must be positive')
+
+            # y is going to be encoded by zero arrays
```
I think this is unnecessary, in the sense that the reader should be aware of what LabelBinarizer is for, or can find out

cperales replied:
Comments removed

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +114,42 @@ def fit(self, X, y=None, sample_weight=None):
+            if len(self.classes_) == 2:
+                ravel = True
+        else:
+            raise ValueError('BaseKernelRidge cannot be used as'
```
If you use python's abc library, the class won't be instantiable

cperales replied:
I think I don't understand pretty much abc library, but I've implemented it ( 5b6d838 ). However, `_BaseKernelRidge` was able to be instanced. Because I tested in a ipython console:

```python
>>> from sklearn.kernel_ridge.kernel_ridge import _BaseKernelRidge
>>> a = _BaseKernelRidge()
>>> a
_BaseKernelRidge(alpha=1, coef0=1, degree=3, gamma=None, kernel='linear',
         kernel_params=None)
```
so `_BaseKernelRidge` can be instanced. What I added was that this class have the property `_estimator_type = None`, so now 'if/elif/else'  makes sense. Before, because `_estimator_type` was not defined in `_BaseKernelRidge`, an error was raised.

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,132 @@
+                             coef0=coef0,
+                             kernel_params=kernel_params)
+
+    def decision_function(self, X):
```
To avoid repeating the implementation, you should call super(...).predict here, and use method overloading to define predict in KRC

cperales replied:
Done e12d1e3

cperales replied:
Done ( e12d1e38daba4ee92b5bbb2c10d431b934d82b8d )

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,105 @@
+
+    clf = KernelRidgeClassifier()
+    Y2 = Y[:-1]  # wrong dimensions for labels
+    assert_raises(ValueError, clf.fit, X, Y2)
```
Tested in common tests (I hope!)

cperales replied:
Done 375f975

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,105 @@
+    Y2 = Y[:-1]  # wrong dimensions for labels
+    assert_raises(ValueError, clf.fit, X, Y2)
+
+    # error for precomputed kernels
```
Why? Because X is the wrong shape? It would be better if you check that the correct error message appears. See other test files. On the other hand check_estimator(KernelRidgeClassifier (kernel='precomputed')) would suffice for this assertion and the next two

cperales replied:
As you said in other comments, that is tested in common test, so that test is removed from test file

jnothman commented on sklearn/kernel_ridge/tests/test_krc.py:
```diff
@@ -0,0 +1,105 @@
+def test_krc_clone_with_callable_kernel():
+    # create KernelRidgeClassifier with callable linear kernel,
+    # check that results are the same as with built-in linear kernel
+    KernelRidgeClassifier_callable = \
```
Don't start local variable names with caps. Here just call it krc_callable

cperales replied:
Done ( da63e60 )

jnothman commented on sklearn/kernel_ridge/tests/test_krr.py:
```diff
@@ -16,12 +14,6 @@
 Y = np.array([y, y]).T
 
 
-def test_kernel_ridge():
```
Why remove this and other tests here? This seems to be the fundamental test of KernelRidge..

cperales replied:
Wow, sorry. Because you said most of these tests were tested in `sklearn/tests/test_common.py`, I tested with coverage which tests were useful or not. I run

```bash
pytest sklearn/kernel_ridge/ sklearn/tests/ --showlocals -v --cov=sklearn/kernel_ridge --cov-report=html:coverage
```

And check the coverage. Then, I removed tests that I thought they were superfluous, and run pytest again. If the number of covered lines hadn't decreased, I reached to the conclusion these tests were superfluous.

However, if you think those tests were essential, I'll add them

cperales replied:
But you are right, `test_kernel_ridge` was important, I added again 5a3bedb

Commit:
Comment removed

Commit:
Sparse accepted

Commit:
_BaseKernelRidge instanced from ABCMeta and BaseEstimator.
Property _estimator_type was added to _BaseKernelRidge as None

Commit:
decision_function for KernelRidgeClassifier is the same as predict for KernelRidge for regresions

Commit:
alpha can be >=0, but not negative, due to Kernel Ridge come from a bound minimization problem
Also, tweak_parameters is removed as @jnothman proposed https://github.com/scikit-learn/scikit-learn/pull/10633/files/dc71c47642ec3768df2445a4eec10e1912cb5ac8#r169183800

sklearn-lgtm commented:
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-9753c5d12a3f5f789eb005ec7e973996504bd4d7)

**new alerts:**

* 2 for Conflicting attributes in base classes

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
PEP8 corrections

Commit:
Cap names for test_krc_clone changed.
Authors added properly

Commit:
Changes for @sklearn-lgtm

Commit:
Test comparing KernelRidge with linear kernel and Ridge, as @jnothman suggested
https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r169184554

cperales reviewed (commented):
Suggestions by @jnothman were taken into consideration and applied

Commit:
PEP8 corrections to 5a3bedb

Commit:
Minor change in a test

cperales commented:
OK, I made a slight change, tests ran again and tests were passed

cperales commented:
By the way, I replied via Github website to all the reviews @jnothman did,
but I still get a denial cross. Does that mean there is another review I
haven't done? Or do I just have to wait until a reviewer check the pull
request again?

I'm finding out that I know too little about open source projects... But
I'm understanding more, step by step. I think this is a good experiencie


Un saludo,
Carlos Perales.

2018-02-14 10:35 GMT+01:00 Joel Nothman <notifications@github.com>:

> *@jnothman* commented on this pull request.
>
> Thanks for the PR. You should expect review to take a while though I admit
> that the estimator does not include a lot of new code, so perhaps try to
> share code with the existing tests (e.g. using pytest.mark.parametrize to
> run on both KRR and KRC) to minimise the workload for reviewers.
>
> Your examples files are named inappropriately now.
>
> We still need just KernelRidge to be the regressor (and not KRR and not
> KRC and not KernelRidgeRegressor)
>
> KernelRidgeClassifier needs to be added to doc/modules/classes.rst
>
> Conventionally, a module which shares a base implementation with other
> modules is called base.py, and the class Base*. Please try to follow such
> conventions.
> ------------------------------
>
> In sklearn/kernel_ridge/kernel_ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168114734>
> :
>
> > +            ravel = False
> +            if len(y.shape) == 1:
> +                y = y.reshape(-1, 1)
> +                ravel = True
> +
> +            self.dual_coef_ = _solve_cholesky_kernel(K, y, alpha,
> +                                                     sample_weight,
> +                                                     copy)
> +            if ravel:
> +                self.dual_coef_ = self.dual_coef_.ravel()
> +
> +        else:
> +            # Check alpha
> +            if self.alpha <= 0:
> +                raise ValueError('C must be positive')
> +            X, y = check_X_y(X, y)
>
> so multi-output isn't an option?
> ------------------------------
>
> In sklearn/kernel_ridge/kernel_ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168115038>
> :
>
> > +            if ravel:
> +                self.dual_coef_ = self.dual_coef_.ravel()
> +
> +        else:
> +            # Check alpha
> +            if self.alpha <= 0:
> +                raise ValueError('C must be positive')
> +            X, y = check_X_y(X, y)
> +            self.classes_ = unique_labels(y)
> +
> +            # y must be encoded by zero arrays with 1 in a position
> +            # assigned to a label
> +            n = X.shape[0]
> +            if self.classes_.shape[0] == 1:
> +                self.n_classes_ = int(self.classes_[0] + 1)
> +                self.classes_ = np.arange(self.n_classes_)
>
> self.classes_ needs to encode all that was in y in the input. It's your
> job as the estimator developer to convert back to that, not to alter what
> the user provides.
> ------------------------------
>
> In sklearn/kernel_ridge/kernel_ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168115912>
> :
>
> > +            if self.alpha <= 0:
> +                raise ValueError('C must be positive')
> +            X, y = check_X_y(X, y)
> +            self.classes_ = unique_labels(y)
> +
> +            # y must be encoded by zero arrays with 1 in a position
> +            # assigned to a label
> +            n = X.shape[0]
> +            if self.classes_.shape[0] == 1:
> +                self.n_classes_ = int(self.classes_[0] + 1)
> +                self.classes_ = np.arange(self.n_classes_)
> +            else:
> +                self.n_classes_ = len(self.classes_)
> +
> +            T = np.zeros((n, self.n_classes_), dtype=np.float64)
> +            # It is essential in order to adapt it to string labels
>
> Why don't you use a LabelEncoder like other classifiers do?
> ------------------------------
>
> In sklearn/kernel_ridge/kernel_ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168116166>
> :
>
> > +
> +    def _get_kernel(self, X, Y=None):
> +        if callable(self.kernel):
> +            params = self.kernel_params or {}
> +        else:
> +            params = {"gamma": self.gamma,
> +                      "degree": self.degree,
> +                      "coef0": self.coef0}
> +        return pairwise_kernels(X, Y, metric=self.kernel,
> +                                filter_params=True, **params)
> +
> +    @property
> +    def _pairwise(self):
> +        return self.kernel == "precomputed"
> +
> +    def fit(self, X, y=None, sample_weight=None):
>
> not a lot of fit seems to be shared between the regressor and classifier.
> why not just put each fit into its corresponding class.
> ------------------------------
>
> In sklearn/kernel_ridge/tests/test_krc.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168118038>
> :
>
> > +    assert_array_equal(krc_cloned.predict(iris.data),
> +                       krc_builtin.predict(iris.data))
> +    assert_array_almost_equal(krc_cloned.predict(iris.data),
> +                              krc_builtin.predict(iris.data))
> +
> +
> +def test_krc_bad_kernel():
> +    krc = KRC(kernel=lambda x, y: x)
> +    assert_raises(ValueError, krc.fit, X, Y)
> +
> +
> +def test_unfitted():
> +    X = "foo!"  # input validation not required when KRC is not fitted
> +
> +    clf = KRC()
> +    assert_raises_regexp(Exception, r"not\b.*\bfitted\b",
>
> This kind of thing is tested by common tests. You do not need to.
> ------------------------------
>
> In sklearn/kernel_ridge/tests/test_krc.py
> <https://github.com/scikit-learn/scikit-learn/pull/10633#discussion_r168118151>
> :
>
> > +
> +def test_krc_parameters():
> +    # Test parameters.
> +    clf = KRC(kernel='linear').fit(X, Y)
> +    assert_array_equal(clf.predict(X), Y)
> +
> +
> +def test_iris():
> +    # Check consistency on dataset iris.
> +
> +    # shuffle the dataset so that labels are not ordered
> +    for k in ('linear', 'rbf'):
> +        clf = KRC(kernel=k).fit(iris.data, iris.target)
> +        assert_greater(np.mean(clf.predict(iris.data) == iris.target), 0.75)
> +
> +    assert_array_equal(clf.classes_, np.sort(clf.classes_))
>
> This kind of thing is tested by common tests. You do not need to.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10633#pullrequestreview-96427620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQiIimnuvwc55aasRLZTymah5VXmYzDQks5tUqjLgaJpZM4SE8E5>
> .
>

jnothman commented:
I've not checked you've met my requests but I'm happy to remove the x. You will still need two approvals

cperales commented:
*While this pull request is being verified for approval, I'd like to give more mathematical background to the contribution. There are some equations, so I recommend to visit this message on Github website https://github.com/scikit-learn/scikit-learn/pull/10633*

Ridge Regression (and therefore this pull request, Ridge Classification) is derived from the method of [least squares](https://en.wikipedia.org/wiki/Least_squares) error for data fitting and a mathematical concept called [Regularization](https://en.wikipedia.org/wiki/Regularization_(mathematics)). To a problem of  error minimization 

![imagen](https://user-images.githubusercontent.com/17336458/36843517-7c5fab36-1d4f-11e8-80d3-dedf0efd32a6.png)

(where *V* is the loss function) is added a regularization term, with a positive parameter alpha

![imagen](https://user-images.githubusercontent.com/17336458/36843550-9803d8bc-1d4f-11e8-80cb-012a3647e5bd.png)

to avoid ill-posed problems. In machine learning, the exact and unique solution usually does not exist, so  ill-posed problems are common. Regularization is then an interesting technique to use which also it helps to keep condition number low.

When the loss function is the L2-norm and *f* is a linear function which assign the label to *x* data,

![imagen](https://user-images.githubusercontent.com/17336458/36843652-dca700c0-1d4f-11e8-92e0-63ef254b476b.png)

there is a specific regularization called [Tikhonov regularization](https://en.wikipedia.org/wiki/Tikhonov_regularization), and Ridge Regression comes from here. If data *X* has an exact and linear relation with target *y* through the matrix *w*, it can be expressed as:

![imagen](https://user-images.githubusercontent.com/17336458/36843667-ec285332-1d4f-11e8-9c7e-2a9b3bed6fce.png)

But, if there isn't any exact solution (it is an ill-posed problem), [Tikhonov regularization](https://en.wikipedia.org/wiki/Tikhonov_regularization) can be used for searching the most similar *w* that solves the problem

![imagen](https://user-images.githubusercontent.com/17336458/36843702-04988478-1d50-11e8-9a63-1011b47230dd.png)

The learning problem with the least squares loss function and Tikhonov regularization can be solved analytically. And kernel trick can be added to the solution in order to attack non-linear problems. Both things are explained detailed [in this paper](http://dro.deakin.edu.au/eserv/DU:30044591/venkatesh-facerecognition-2007.pdf). For linear problems (without kernel trick), analytical solution is

![imagen](https://user-images.githubusercontent.com/17336458/36844533-67ade3b6-1d53-11e8-8663-4911a33ca54c.png)


Function *f*, which assigns label to data *x*, is expressed with kernel trick as

![imagen](https://user-images.githubusercontent.com/17336458/36844553-7b993efc-1d53-11e8-8064-b98507fa90d2.png)


Where *k* is the kernel trick function.

This pull request extends kernel ridge regression to classification, mainly by binarizing the label *y* with `sklearn.preprocessing.LabelBinarize`. Also, tests and documentation are added.

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,144 @@
+# Author: Carlos Perales <sir.perales@gmail.com>
```
Fwiw, splitting this into separate files provides no advantage imo, but means you often need to edit two or three files at once to make a docstring change. I know we've done such splitting elsewhere in the library, and inconsistently so,  but often not when there is so much shared by the implementations

cperales replied:
Thanks for the info. Using that Python does not force you to put every object in a separated file, I could move the three objects (base and the inherits ones) into an unique file. Do you prefer so? Or would it break some conventions?

However, as it is now, I think it does not provide so much problems.

jnothman replied:
Personally I'd appreciate them all staying in kernel_ridge.py. That seems much easier to maintain since so much of the code is shared.

cperales replied:
Done ( 39fdc7aa35169112b4f16c8f9034283abc02e55f ). What's more, I think then no longer a specific folder for `kernel_ridge` is needed.

jnothman replied:
Yes, I had been trying to suggest that a folder was unnecessary

cperales replied:
Perfect then, sorry for not understanding you before

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -2,39 +2,23 @@
 
```
The convention is to call such a file base. Py

cperales replied:
Done, renamed from `kernel_ridge.py` to `base_ridge.py` 888c59e

jnothman commented on examples/kernel_ridge/plot_krc_custom_kernel.py:
```diff
@@ -0,0 +1,61 @@
+
+# Plot also the training points
+plt.scatter(X[:, 0], X[:, 1], c=Y, cmap=plt.cm.Paired, edgecolors='k')
+plt.title('3-Class classification using Support Vector Machine with custom'
```
More generally I feel like this might be an unnecessary proliferation of examples, which may make it harder for users to find the information they need.

cperales replied:
I see... so, linking with with [main review comment](https://github.com/scikit-learn/scikit-learn/pull/10633#pullrequestreview-100589899), you suggest to unify SVM and Ridge plot and examples, right?

Personally, I don't think that is a good idea, since every test and example is organized by folders, so many folders would be removed and/or merged. Besides, unifying SVR/SVC (which calls to `libsvm`in C) and Kernel Ridge/Classifier could be caotic since the implementation in scikit-learn is different.

cperales replied:
On one hand, **SVM** narrative and minimization **problem is geometrical**, which does not allow multiclassification. Besides, SVM thinks about *how to build an hyperplane that separates class 1 from class 2*, and different tricks are applied in order to make it multiclass and also available for regressions.

On the other hand, **Ridge** narrative and minimization **problem is analytical**, because it tries to look for the *w* that 

![image](https://user-images.githubusercontent.com/17336458/36843667-ec285332-1d4f-11e8-9c7e-2a9b3bed6fce.png)

using Tikhonov regularization directly in the minimization problem. 

So they either does not start from the same hypothesis or achieve to the same solution. That's why I've introduced Kernel Ridge Classifier in the documentation expanding Kernel Ridge ( https://github.com/scikit-learn/scikit-learn/pull/10633/files#diff-242cc346ae920fe5058a877fbae888e8 )

jnothman replied:
Basically I don't know that we need more examples. What we need is to make clear in the narrative documentation (`doc/` rather than `examples/`) that we have alternative kernel-based classifiers available. That documentation, and/or an example, should help a user decide (a) whether a kernel classifier is a good choice for them; (b) whether SVC or KRC is a better choice for them. Beyond that, how does an extended example help?

cperales replied:
Ok, 00b85b6be0bce0850ac19907d1f01a580ec5f0dc and 23a7e37 (new commit) are aimed to provide documentation as you requested

jnothman replied:
So what of the redundant example?

jnothman commented on sklearn/kernel_ridge/kernel_ridge.py:
```diff
@@ -143,23 +118,39 @@ def fit(self, X, y=None, sample_weight=None):
-                         y_numeric=True)
         if sample_weight is not None and not isinstance(sample_weight, float):
             sample_weight = check_array(sample_weight, ensure_2d=False)
+        # Check alpha
```
Rm needless comment

cperales replied:
Done 888c59e

jnothman reviewed (commented):
I've not reviewed implementation and tests this time. You need a see also from SVC and perhaps also from its narrative docs, as I think this is basically the same hypothesis space with a different loss

Commit:
Removed needless comment
renamed kernel_ridge.py to base_ridge.py

jnothman commented on sklearn/kernel_ridge/krc.py:
```diff
@@ -0,0 +1,144 @@
+
+    classes_ : array of shape = [n_classes] or a list of such arrays
+        The classes labels (single output problem),
+        or a list of arrays of class labels (multi-output problem).
```
I don't think we support multi-output classification here

cperales replied:
You are right, I'll remove it

jnothman reviewed (commented):
Please add a test that asserts the same coef_ for corresponding KernelRidge and KernelRidgeClassifier problems.

cperales commented:
About the [coef_ assertion you said](https://github.com/scikit-learn/scikit-learn/pull/10633#pullrequestreview-100991568), how should the comparison be done? For regression, coef_ is a vector (a matrix *m x 1*, where *m* is the number of features) and for classification coef_ is a matrix (*m x t*, where *t* is the number of classes to classify)

Commit:
Kernel ridge objects are moved into one file, so a folder is not longer necessary

Some imports were fixed

Commit:
More documentation, comparing SVM with Kernel Ridge

jnothman commented:
KernelRidge supports multiple regression, which should produce a cord matrix

Commit:
Avoid further calculations if X, y are not checked

Commit:
This test asserts that KernelRidge and KernelRidgeClassifier have the same dual_coef_ when classification target are specified as regressions

cperales commented:
Now I get [what @jnothman said](https://github.com/scikit-learn/scikit-learn/pull/10633#pullrequestreview-100991568). So I introduced ( 1d1bc0d ) a test that, on one hand, calls KernelRidgeClassifier for a classification problem; on th other hand, target are binarized and transformed into a regression problem. Then, a dual_coef_ comparison asserts that they achieve to the same dual_coef_ solution.

jnothman commented:
great, thanks. sorry I'm not sure when I'll afford this a full review, but
I'm glad it's getting closer to being mergeable

cperales commented:
Hello there, @jnothman and all!

This pull request has been quiet for a long time. Is there something I can do?

I truly believe that the contribution this pull request brings is very simple. I even just used the functions and objects scikit-learn already has to write the code of this classifier.

If you think it needs more documentation, or to bring a specific point to the documentation, tell me how and where I should write it.

jnothman commented:
> This pull request has been quiet for a long time.

I opened this conversation with "Thanks for the PR. You should expect review to take a while".

Four weeks is far from long. Try pinging around June or July if this continues in silence.

cperales commented:
OK then. I'm still offering more documentation if it's necessary ;)

El lun., 2 abr. 2018 15:04, Joel Nothman <notifications@github.com>
escribió:

> This pull request has been quiet for a long time.
>
> I opened this conversation with "Thanks for the PR. You should expect
> review to take a while".
>
> Four weeks is far from long. Try pinging around June or July if this
> continues in silence.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10633#issuecomment-377924741>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQiIiplJzLQiliFbeYCq3kLHRQYtQWFsks5tkiHVgaJpZM4SE8E5>
> .
>

cperales commented:
I'd like to ask again about this pull request @jnothman . Is it still be considered? Can I help somehow, maybe with documentation about some parts?

jnothman commented:
I might be able to look at this in a few days. That doesn't mean it will be merged for the upcoming release!

cperales commented:
Ok! I just want this to move on :) I still believe it is a great method and **scikit-learn** should have it implemented

cperales commented:
Hi @jnothman ! I'm pretty sure you have much other pull request to review, and most of them will fix some currently bugs or enhancements proposed... but I'm still interested in this pull request.

Do you need me to help you with this pull request? I'll be able to add more documentation or even a restructuration if you feel it is necessary. Perhaps @amueller might have a harsher critique of this proposal

jnothman commented on doc/modules/kernel_ridge.rst:
```diff
@@ -50,8 +50,17 @@ prediction time depends on the parameters :math:`\epsilon` and :math:`C` of the
 
+Following same mathematical formulation, with
+a label encoding Kernel Ridge can be used for classification
+[V2007]_. :class:`` uses same logic as
```
missing reference

jnothman commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -1,85 +1,143 @@
-    assert_array_almost_equal(K, K2)
-
-
-def test_kernel_ridge_sample_weights():
```
why was this removed?

cperales replied:
It is not removed; in the same file (_test_kernel_ridge.py_), it appears some tests for classification (_lines from 43 to 86_)  and for regression (_lines from 89 to 143_).

jnothman commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -1,85 +1,143 @@
     assert_array_almost_equal(pred, pred2)
 
 
-def test_kernel_ridge_precomputed():
```
why was this removed?

cperales replied:
It is not removed; in the same file (_test_kernel_ridge.py_), it appears some tests for classification (_lines from 43 to 86_)  and for regression (_lines from 89 to 143_).

jnothman commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -1,85 +1,143 @@
-        assert_array_almost_equal(pred, pred2)
-
-
-def test_kernel_ridge_precomputed_kernel_unchanged():
```
why was this removed?

cperales replied:
It is not removed; in the same file (_test_kernel_ridge.py_), it appears some tests for classification (_lines from 43 to 86_)  and for regression (_lines from 89 to 143_).

jnothman commented on sklearn/kernel_ridge.py:
```diff
@@ -183,3 +173,234 @@ def predict(self, X):
+    """Kernel Ridge Classifier.
+
+    Kernel Ridge Classifier (KRC) combines ridge
+    regression (linear least squares with l2-norm regularization), special
```
in what way is it "special"?

jnothman replied:
"In what way is it special" is a reference to this line. What you mean by "special encoding for target"?

jnothman commented on doc/modules/kernel_approximation.rst:
```diff
@@ -4,11 +4,26 @@ Kernel Approximation
-The following feature functions perform non-linear transformations of the
-input, which can serve as a basis for linear classification or other
-algorithms.
+correspond to certain kernels, as they are used, for example, by
```
This section on kernel approximation shouldn't really be discussing ridge vs svm, which tend to use exact kernels.

cperales replied:
Solved in ff40a1dc

jnothman reviewed (commented):
You should mention the classifier in the `kernel_ridge` section of the user guide.

jnothman commented:
some quick points, sorry for not more attention

Commit:
Merge branch 'master' into pull_dev

Commit:
Correction to https://github.com/scikit-learn/scikit-learn/pull/10633/files/1d1bc0dde9860f77d23a6ae31e8acc38617bb444#r223212477

cperales commented:
OK, i'll review your changes and i'll apply the relevant changes between today and tomorrow

cperales reviewed (commented):
Review done! @jnothman

jnothman commented:
Yes, master was failing due to an inappropriate merge yesterday. Should now
be fixed.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

cperales commented:
It is! Now tests passes

jnothman commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -1,85 +1,143 @@
+
+# Tests for classification
+def test_unicode_kernel():
+    # Test that a unicode kernel name does not cause a TypeError
```
Python 2 is no longer supported. This is no longer needed.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -70,8 +70,9 @@
                 'GaussianProcessRegressor', 'TransformedTargetRegressor',
-                'KNeighborsRegressor', 'KernelRidge', 'Lars', 'Lasso',
-                'LassoLars', 'LinearRegression', 'MultiTaskElasticNet',
+                'KNeighborsRegressor', 'KernelRidge',
```
I don't think there's any change here. Please revert changes to the file.

jnothman reviewed (commented):
Except for redundant examples, this LGTM.

cperales commented:
any news?

amueller commented:
@cperales have you addressed @jnothman's comments above? Also it's waiting for a second reviewer.

ogrisel commented:
@cperales why did you close this PR? No more time to work on this?

For people interested in this feature, we could revive this PR and make it up to date w.r.t. input validation check API.

In the mean time, here is a minimal implementation that should work for most use cases (if you ignore the class docstring that would need to be overridden.

```python
from sklearn.base import ClassifierMixin
from sklearn.kernel_ridge import KernelRidge
from sklearn.preprocessing import LabelBinarizer


class KernelRidgeClassifier(ClassifierMixin, KernelRidge):

    def fit(self, X, y):
        self.lb_ = LabelBinarizer()
        Y = self.lb_.fit_transform(y)
        self.classes_ = self.lb_.classes_
        return super().fit(X, Y)

    def predict(self, X):
        return self.lb_.classes_[super().predict(X).argmax(axis=1)]

```

ogrisel commented:
Actually for consistency with the linear case, one might want to prefer to use `LabelBinarizer(pos_label=1, neg_label=-1)` as done in this PR.

cperales commented:
Hi @ogrisel . After almost a year after amueller comment ( https://github.com/scikit-learn/scikit-learn/pull/10633#issuecomment-518841134 ), I understood that other reviewer was needed. But he/she never came, so I canceled the pull request. But I still find it interesting, I can reopen it.

It seems I couldn't reopen easily, as the screenshot I attach says:

![image](https://user-images.githubusercontent.com/17336458/114376904-60fc4300-9b86-11eb-9a85-900bb1a1b43d.png)


But I can reimplement, it is not a big deal.

ogrisel commented:
@cperales the (partial) second review https://github.com/scikit-learn/scikit-learn/pull/10633#pullrequestreview-193057649 above was never addressed, hence the confusion.

ogrisel commented on doc/modules/svm.rst:
```diff
@@ -35,6 +35,10 @@ The disadvantages of support vector machines include:
 
+    - SVMs do not directly attack multiclass problems, just binary classification.
+      Instead, one-against-all and one-against-one techniques are applied
+      for multiclass classification.
```
Maybe mention `KernelRidgeClassifier` explicitly here as naturally multiclass kernel-based classifier here.

cperales reviewed (commented):
Okay then, I'll try to reimplement this code. However, I can't reopen this branch, as I show in a screenshot in answer https://github.com/scikit-learn/scikit-learn/pull/10633#issuecomment-817671358

How should I proceed? I think it would be easier if I just make another pull request, and I reference this closed pull request (since I can't reopen it) and I mention you @ogrisel . Do you agree?

cperales commented:
Okay then, I'll try to reimplement this code. However, I can't reopen this branch, as I show in a screenshot in answer https://github.com/scikit-learn/scikit-learn/pull/10633#issuecomment-817671358

How should I proceed? I think it would be easier if I just make another pull request, and I reference this closed pull request (since I can't reopen it) and I mention you @ogrisel . Do you agree?

ogrisel commented on examples/kernel_ridge/plot_krc_rbf_parameters.py:
```diff
@@ -0,0 +1,173 @@
+The ``C`` parameter trades off misclassification of training examples against
+simplicity of the decision surface. A low ``C`` makes the decision surface
+smooth, while a high ``C`` aims at classifying all training examples correctly
+by giving the model freedom to select more samples as support vectors.
```
As our Ridge models are parametrized in terms of alpha instead of C this paragraph will need to be rewritten are `alpha` is inversely proportional to `C`.

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    alpha : {float, array-like}, shape = [n_targets]
+        Small positive values of alpha improve the conditioning of the problem
+        and reduce the variance of the estimates.  Alpha corresponds to
+        ``(2*C)^-1`` in other linear models such as LogisticRegression or
```
I think there is a `n_samples` (or equivalently `sample_weight.sum()` when `sample_weight` is not None) that appears in the relation between Ridge's `alpha` and SVM or LogisticRegression `C` parameter (to be checked).

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    kernel : string or callable, default="linear"
+        Kernel mapping used internally. A callable should accept two arguments
+        and the keyword arguments passed to this object as kernel_params, and
+        should return a floating point number.
```
I think we should tell the user that `KernelRidgeClassifier(kenel="linear")` is equivalent to `RidgeClassifier()` which is typically preferable for the linear case.

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    ----------
+    * Kevin P. Murphy
+      "Machine Learning: A Probabilistic Perspective", The MIT Press
+      chapter 14.4.3, pp. 492-493
```
Please add a link to the online page of this reference and follow the formatting of this docstring for instance:

https://github.com/scikit-learn/scikit-learn/blob/95119c13a/sklearn/metrics/cluster/_unsupervised.py#L97-L106

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    >>> y = rng.randn(n_samples)
+    >>> X = rng.randn(n_samples, n_features)
+    >>> clf = KernelRidge(alpha=1.0)
+    >>> clf.fit(X, y) # doctest: +NORMALIZE_WHITESPACE
```
`# doctest: +NORMALIZE_WHITESPACE` is no longer needed as it is configured globally, please remove.

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    sklearn.svm.SVR:
+        Support Vector Regression implemented using libsvm.
+    sklearn.kernel_ridge.KernelRidgeClassifier:
+        Kernel Ridge implemented for multiclass clasifications.
```
typo: classification

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+        and reduce the variance of the estimates.  Alpha corresponds to
+        ``(2*C)^-1`` in other linear models such as LogisticRegression or
+        LinearSVC. If an array is passed, penalties are assumed to be specific
+        to the targets. Hence they must correspond in number.
```
Maybe add a generic sentence such as:

High values of alpha will tend to make the learned decision function smoother but potentially cause underfitting while smaller values of alpha will preserve the influence of the individual training datapoints and therefore overfitting although this highly depends on the choice of the kernel parameters.

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+    * An, Senjian, Wanquan Liu, and Svetha Venkatesh.
+      "Face recognition using kernel ridge regression." Computer Vision
+      and Pattern Recognition, 2007. CVPR'07. IEEE Conference on. IEEE,
+      2007.
```
Same comment as previously about formatting and linking references.

ogrisel commented on sklearn/kernel_ridge.py:
```diff
@@ -192,3 +182,234 @@ def predict(self, X):
+            case, confidence score for self.classes_[1] where >0 means this
+            class would be predicted.
+        """
+        return super(KernelRidgeClassifier, self).predict(X=X)
```
I think in Python3 we can just do `return super().predict(X)`. And the `X=X` is weird for a method with a single mandatory positional arg.

ogrisel commented on sklearn/tests/test_kernel_ridge.py:
```diff
@@ -1,85 +1,143 @@
+    Y_krr = label_binarize(Y, classes=np.array([1, 2]),
+                           neg_label=-1, pos_label=1).ravel()
+    krr = KernelRidge(kernel='linear', alpha=1).fit(X, Y_krr)
+    assert_array_equal(krc.dual_coef_, krr.dual_coef_)
```
Good sanity check. Could also please add a consistency check that `KernelRidgeClassifier(kernel='linear')` and `RidgeClassifier()` converge to the same decision function, for instance by checking that they have exactly the same hard predictions and `decision_function` ouputs on a random test set?

Could you also please add a test to check that `KernelRidgeClassifier` can reach a 100% classification accuracy on toy a non-linearly separable problem (for instance make_circle or make_swiss_roll)?

ogrisel reviewed (commented):
Here are more review comments to address:

ogrisel commented:
> How should I proceed? I think it would be easier if I just make another pull request, and I reference this closed pull request (since I can't reopen it) and I mention you @ogrisel . Do you agree?

Indeed, I cannot reopen it either because of the master branch renaming. Please open a new PR then as you suggested.
